### PR TITLE
feat(size-report): add proper size report check

### LINKED ISSUE #2634: Build per-component report for CSS size
Create simple HTML table &/or leave PR comment with bundle sizes.  Maybe fail the build based on large increase in size?

### PR DESCRIPTION
**Pull request by redallen**
Closes #2634

- Adds proper size reporting vs version we cut branch off of from NPM.
  - Fails if file grows > 50% or `dist/*.css` grows > 10%
- Leave PR comment after every reporting step (so one for preview, a11y, and CSS size report)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -156,5 +156,8 @@
     "hooks": {
       "commit-msg": "commitlint -E HUSKY_GIT_PARAMS"
     }
+  },
+  "dependencies": {
+    "@octokit/rest": "^16.40.1"
```
Comment (redallen) [inline package.json]: Move to dev dependencies

File: .circleci/css-size-report/sizeReport.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+let html = '<table>';
+html += '<tr>';
+html += `<td>Name</td>`;
+html += `<td>Current(kb)</td>`;
+html += `<td>Previous(kb)</td>`;
+html += `<td>Diff %</td>`;
+html += '</tr>';
+
+// @todo add: 'dist/patternfly.css', 'dist/patternfly.min.css'
```
Comment (redallen) [inline .circleci/css-size-report/sizeReport.js]: Do the todo!

File: .circleci/css-size-report/sizeReport.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+glob
+  .sync('../../dist/components/**/*.css')
```
Comment (redallen) [inline .circleci/css-size-report/sizeReport.js]: We can we include layouts too? We have one duplicate (Flex layout vs Flex component), so you'd have to handle that special case in the HTML table.

File: .circleci/css-size-report/sizeReport.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  // .sort(by size)
+  .forEach(({ file, size }) => {
+    if (size >= 10000) {
+      sizeCol = '#E74C3C';
+    } else if (size > 8000 && size < 10000) {
+      sizeCol = '#F1C40F';
+    } else {
+      sizeCol = '#229954';
+    }
+    prevResult = glob
```
Comment (redallen) [inline .circleci/css-size-report/sizeReport.js]: Move this outside of the loop and do `const prevSizes = glob.sync('node_modules/@patternfly/patternfly/**/*.css')...`

File: .circleci/config.yml [hunk 5]
```diff
@@ -12,6 +12,12 @@ references:
   - src/icons/PfIcons/
   js_deps_cache_key: &js_deps_cache_key
     js-deps-v{{.Environment.CACHE_VERSION}}-{{checksum "package-lock.json"}}
+  size_report_paths: &size_report_paths
```
Comment (redallen) [inline .circleci/config.yml]: We can revert all these lines added.

File: .circleci/config.yml [hunk 6]
```diff
@@ -108,12 +114,24 @@ jobs:
...
         command: npm run lint:js
-    - run:
-        name: Lint CSS Size
-        command: npm run lint:size
+        when: always
     - run:
         name: Lint CSS
         command: npm run lint:css
+        when: always
+    - restore_cache:
```
Comment (redallen) [inline .circleci/config.yml]: No need to restore_cache

File: .circleci/config.yml [hunk 7]
```diff
@@ -108,12 +114,24 @@ jobs:
...
+        keys:
+        - *size_report_master_cache_key
+    - run:
+        name: Lint CSS Size
+        command: .circleci/css-size-report/css-size-report.sh
+        when: always
+    - persist_to_workspace:
+        root: ~/project
+        paths: *size_report_paths
+    - save_cache:
```
Comment (redallen) [inline .circleci/config.yml]: No need to save_cache

File: package.json (hunk 1 above)
Comment (ddonahue007) [reply package.json]: Done

File: .circleci/config.yml (hunk 5 above)
Comment (ddonahue007) [reply .circleci/config.yml]: done

File: .circleci/config.yml (hunk 6 above)
Comment (ddonahue007) [reply .circleci/config.yml]: done

File: .circleci/config.yml (hunk 7 above)
Comment (ddonahue007) [reply .circleci/config.yml]: done

File: .circleci/css-size-report/sizeReport.js (hunk 4 above)
Comment (ddonahue007) [reply .circleci/css-size-report/sizeReport.js]: Done

File: .circleci/config.yml (hunk 6 above)
Comment (ddonahue007) [reply .circleci/config.yml]: I removed restore cache for the below. You noted above the 'Lint CSS' target.
    - run:
        name: Lint CSS Size
        command: .circleci/css-size-report/css-size-report.sh
        when: always

File: .circleci/css-size-report/sizeReport.js (hunk 2 above)
Comment (ddonahue007) [reply .circleci/css-size-report/sizeReport.js]: Done

File: .circleci/css-size-report/sizeReport.js (hunk 3 above)
Comment (ddonahue007) [reply .circleci/css-size-report/sizeReport.js]: DOne

**Comment by ddonahue007** [comment]
Screen Shot 2020-02-07 at 12 28 27 PM

**Comment by ddonahue007** [comment]
<!DOCTYPE html><body><table style="width:100%"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table></body></html>

**Comment by ddonahue007** [comment]
<!DOCTYPE html><body><table style="width:100%"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table></body></html>

**Comment by ddonahue007** [comment]
<!DOCTYPE html><body><table style="width:100%"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table></body></html>

**Comment by ddonahue007** [comment]
<!DOCTYPE html><body><table style="width:100%"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table></body></html>

**Comment by ddonahue007** [comment]
<table id="css-lint-table"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table>

**Comment by ddonahue007** [comment]
<table id="css-lint-table"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table>

**Comment by ddonahue007** [comment]
<table id="css-lint-table"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table>

**Comment by ddonahue007** [comment]
<table id="css-lint-table"><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>626630</td><td>-6</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>624816</td><td>-6</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>548596</td><td>-6</td><tr></table>

**Comment by ddonahue007** [comment]
<table id='css-lint-size'><tr><th>Name</th><th>Current(kb)</th><th>Previous(kb)</th><th>Diff %</th></tr><tr><td>patternfly.css</td><td>591029</td><td>648059</td><td>-10</td><tr><tr><td>patternfly-no-reset.css</td><td>589215</td><td>646245</td><td>-10</td><tr><tr><td>patternfly.min.css</td><td>518060</td><td>568815</td><td>-10</td><tr><tr><td>components/AboutModalBox/about-modal-box.css</td><td>36868</td><td>13646</td><td>63</td><tr><tr><td>components/Nav/nav.css</td><td>30610</td><td>31016</td><td>-1</td><tr><tr><td>components/Table/table.css</td><td>22203</td><td>22458</td><td>-1</td><tr><tr><td>components/Switch/switch.css</td><td>6715</td><td>6366</td><td>5</td><tr><tr><td>components/SimpleList/simple-list.css</td><td>4735</td><td>4762</td><td>-1</td><tr><tr><td>components/ModalBox/modal-box.css</td><td>4655</td><td>5090</td><td>-9</td><tr><tr><td>components/Pagination/pagination.css</td><td>3786</td><td>4077</td><td>-8</td><tr><tr><td>components/Title/title.css</td><td>2678</td><td>2704</td><td>-1</td><tr></table>

**Pull request merged**